plugins/modules/digital_ocean_snapshot.py

```diff
@@ -0,0 +1,265 @@
+                            json["action"]["resource_id"]
+                        ),
+                    )
+                self.module.exit_json(
+                    changed=True,
+                    msg="Created snapshot, action {0}".format(self.action_id),
+                )
+            else:
+                self.module.fail_json(
+                    changed=False,
+                    msg="Failed to create snapshot: {0}".format(json["message"]),
+                )
+        elif self.snapshot_type == "volume":
+            data = {
+                "name": self.snapshot_name,
+            }
+            response = self.rest.post(
+                "volumes/{0}/snapshots".format(str(self.volume_id)), data=data
+            )
+            status = response.status_code
+            json = response.json
+            if status == 201:
+                self.module.exit_json(
+                    changed=True,
+                    msg="Created snapshot, snapshot {0}".format(json["snapshot"]["id"]),
+                )
+            else:
+                self.module.fail_json(
+                    changed=False,
+                    msg="Failed to create snapshot: {0}".format(json["message"]),
+                )
+
+    def delete(self):
+        response = self.rest.delete("snapshots/{0}".format(str(self.snapshot_id)))
+        status = response.status_code
+        if status == 204:
+            self.module.exit_json(
+                changed=True,
+                msg="Deleted snapshot {0}".format(str(self.snapshot_id)),
+            )
+        else:
+            json = response.json
+            self.module.fail_json(
+                changed=False,
+                msg="Failed to delete snapshot {0}: {1}".format(
+                    self.snapshot_id, json["message"]
+                ),
+            )
+
+
+def run(module):
+    state = module.params.pop("state")
+    snapshot = DOSnapshot(module)
+    if state == "present":
+        snapshot.create()
+    elif state == "absent":
+        snapshot.delete()
+
+
+def main():
+    argument_spec = DigitalOceanHelper.digital_ocean_argument_spec()
+    argument_spec.update(
+        state=dict(choices=["present", "absent"], default="present"),
+        snapshot_type=dict(
+            type="str", required=False, choices=["droplet", "volume"], default="droplet"
+        ),
+        snapshot_name=dict(type="str"),
+        droplet_id=dict(type="str"),
+        volume_id=dict(type="str"),
+        snapshot_id=dict(type="str"),
+        wait=dict(type="bool", default=True),
+        wait_timeout=dict(default=120, type="int"),
+    )
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_if=[
+            ("state", "present", ["snapshot_name"]),
+            ("state", "absent", ["snapshot_id"]),
+            ("snapshot_type", "droplet", ["droplet_id"]),
+            ("snapshot_type", "volume", ["volume_id"]),
```

Comment:
Are those two really needed? I would assume this could cause trouble when trying to delete snapshots because the user would need to provide a droplet ID. I would assume that dropped or volume IDs are only needed when the state is set to present.